lib/platform/git/storage.js

```diff
@@ -0,0 +1,201 @@
+      git = Git(repoDir.path).silent(true);
+      await git.clone(config.url, '.');
+
+      if (config.gitAuthor) {
+        await git.raw(['config', 'user.name', config.gitAuthor.name]);
+        await git.raw(['config', 'user.email', config.gitAuthor.address]);
+        // not supported yet
+        await git.raw(['config', 'commit.gpgsign', 'false']);
+      }
+
+      // see https://stackoverflow.com/a/44750379/1438522
+      config.baseBranch =
+        config.baseBranch ||
+        (await git.raw(['symbolic-ref', 'refs/remotes/origin/HEAD']))
+          .replace('refs/remotes/origin/', '')
+          .trim();
+    }
+
+    async function createBranch(branchName, sha) {
+      await git.reset('hard');
+      await git.checkoutBranch(branchName, sha);
+      await git.push(['origin', branchName, '--force']);
+    }
+
+    // Return the commit SHA for a branch
+    async function getBranchCommit(branchName) {
+      const res = await git.revparse(['origin/' + branchName]);
+      return res.trim();
+    }
+
+    async function getCommitMessages() {
+      logger.debug('getCommitMessages');
+      const res = await git.log({
+        n: 10,
+        format: { message: '%s' },
+      });
+      return res.all.map(commit => commit.message);
+    }
+
+    function setBaseBranch(branchName) {
+      if (branchName) {
+        logger.debug(`Setting baseBranch to ${branchName}`);
+        config.baseBranch = branchName;
+      }
+    }
+
+    async function getFileList(branchName) {
+      const branch = branchName || config.baseBranch;
+      const exists = await branchExists(branch);
+      if (!exists) {
+        return [];
+      }
+      const files = await git.raw([
+        'ls-tree',
+        '-r',
+        '--name-only',
+        'origin/' + branch,
+      ]);
+      return files.split('\n').filter(Boolean);
+    }
+
+    async function branchExists(branchName) {
+      try {
+        await git.raw(['show-branch', 'origin/' + branchName]);
+        return true;
+      } catch (ex) {
+        return false;
+      }
+    }
+
+    async function getAllRenovateBranches(branchPrefix) {
+      const branches = await git.branch(['--remotes', '--verbose']);
+      return branches.all
+        .map(localName)
+        .filter(branchName => branchName.startsWith(branchPrefix));
+    }
+
+    async function isBranchStale(branchName) {
+      const branches = await git.branch([
+        '--remotes',
+        '--verbose',
+        '--contains',
+        config.baseBranch,
+      ]);
+      return !branches.all.map(localName).includes(branchName);
+    }
+
+    async function deleteBranch(branchName) {
+      await git.raw(['push', '--delete', 'origin', branchName]);
+      try {
+        await git.deleteLocalBranch(branchName);
+      } catch (ex) {
+        // local branch may not exist
+      }
+    }
+
+    async function mergeBranch(branchName) {
+      await git.reset('hard');
+      await git.checkoutBranch(branchName, 'origin/' + branchName);
```

Comment:
This fails with `fatal: A branch named 'renovate/node-8.x' already exists` if performed in the same run as the branch was created. OK to just add a try/catch and ignore it? Or should we merge from `origin/` and not check it out locally?

Reply:
this is fixed by using `checkout -B` instead of `checkout -b`, the first resets the branch if already exits, the second throws